c_api/src/android.rs

```diff
@@ -0,0 +1,477 @@
+    let settings_cstr = java_string_to_cstring_jint!(env, settings);
+    let format_cstr = java_string_to_cstring_jint!(env, format);
+    
+    unsafe { c2pa_load_settings(settings_cstr.as_ptr(), format_cstr.as_ptr()) }
+}
+
+/// Returns a ManifestStore JSON string from a file path
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_readFile(
+    mut env: JNIEnv,
+    _class: JClass,
+    path: JString,
+    data_dir: JString,
+) -> jstring {
+    let path_cstr = java_string_to_cstring_jstring!(env, path);
+    let data_dir_cstr = java_string_to_cstring_opt_jstring!(env, data_dir);
+    
+    let data_dir_ptr = data_dir_cstr.as_ref().map_or(ptr::null(), |s| s.as_ptr());
+    let result = unsafe { c2pa_read_file(path_cstr.as_ptr(), data_dir_ptr) };
+    
+    c_string_to_java_string(&mut env, result)
+}
+
+/// Returns an Ingredient JSON string from a file path
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_readIngredientFile(
+    mut env: JNIEnv,
+    _class: JClass,
+    path: JString,
+    data_dir: JString,
+) -> jstring {
+    let path_cstr = java_string_to_cstring_jstring!(env, path);
+    let data_dir_cstr = java_string_to_cstring_jstring!(env, data_dir);
+    
+    let result = unsafe { c2pa_read_ingredient_file(path_cstr.as_ptr(), data_dir_cstr.as_ptr()) };
+    
+    c_string_to_java_string(&mut env, result)
+}
+
+/// Creates a C2paBuilder from a JSON manifest definition string
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_builderFromJson(
+    mut env: JNIEnv,
+    _class: JClass,
+    manifest_json: JString,
+) -> jlong {
+    let manifest_json_cstr = java_string_to_cstring_jlong!(env, manifest_json);
+    
+    let builder = unsafe { c2pa_builder_from_json(manifest_json_cstr.as_ptr()) };
+    builder as jlong
+}
+
+/// Frees a C2paBuilder
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_builderFree(
+    _env: JNIEnv,
+    _class: JClass,
+    builder_ptr: jlong,
+) {
+    if builder_ptr != 0 {
+        unsafe { c2pa_builder_free(builder_ptr as *mut C2paBuilder) };
+    }
+}
+
+/// Sets the no-embed flag on the Builder
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_builderSetNoEmbed(
+    _env: JNIEnv,
+    _class: JClass,
+    builder_ptr: jlong,
+) {
+    if builder_ptr != 0 {
+        unsafe { c2pa_builder_set_no_embed(builder_ptr as *mut C2paBuilder) };
+    }
+}
+
+/// Sets the remote URL on the Builder
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_builderSetRemoteUrl(
+    mut env: JNIEnv,
+    _class: JClass,
+    builder_ptr: jlong,
+    remote_url: JString,
+) -> jint {
+    if builder_ptr == 0 {
+        return -1;
+    }
+    
+    let remote_url_cstr = java_string_to_cstring_jint!(env, remote_url);
+    
+    unsafe { c2pa_builder_set_remote_url(builder_ptr as *mut C2paBuilder, remote_url_cstr.as_ptr()) }
+}
+
+/// Creates and verifies a C2paReader from an asset stream with the given format
+#[no_mangle]
+pub extern "system" fn Java_org_c2pa_C2PA_readerFromStream(
```

Comment:
Here we have the reader - that is similar in spirit to `Java_org_c2pa_C2PA_readFile`, if we open the file as stream and pass it to a Reader, to finally get the manifest store as JSON.